### PR TITLE
compressFirmwareXz: fix links to directories

### LINKED ISSUE #283005: gsp for tu117 is not getting loaded linux-firmware issue
### Describe the bug
I found this while I am making this package request https://github.com/NixOS/nixpkgs/issues/280326
and basically 
these images explain everything
image
image

basically as I said in the oroginal comment: https://github.com/NixOS/nixpkgs/issues/280326#issuecomment-1902063376
I think the problem is that its linking ../tu116/gsp.xz instead of ../tu116/gsp
and since the file system is read only and I am not knowledgable enough I don't really know how to test this theory

### Steps To Reproduce
1. have a card that uses that firmware tu117
2. 
```
   boot.kernelParams = [
      "nouveau.config=NvGspRm=1"
      "nouveau.debug=info,VBIOS=info,gsp=debug"
    ];
```
4. journalctl -b 0

### Expected behavior
the firmware should load

### Notify maintainers
@fpletz 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.1, NixOS, 24.05 (Uakari), 24.05.20240121.612f972`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"home-manager, nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by fpletz**
## Description of changes

Only add the `.xz` suffix to links if link target is a regular file. This breaks if the target is a directory.

Fixes #283005.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by fpletz** [comment]
cc @beh-10257

**Comment by beh-10257** [comment]
Will this get into unstable repos soon enough or is there a way to test this
Hopefully I can take this in the next two days when I ran
Nix flake update
And thanks

**Review by alyssais — APPROVED** [review]

**Comment by beh-10257** [comment]
I know this is already approved but its been two days and this is not merged and I guess it will take another two days
After getting merged to be in nixpkgs-unstable

is this in staging or something
How do I add staging as an input using flakes and use linux-firmware from staging
So I can continue nvk testing
Thanks either way

**Comment by getchoo** [comment]
> is this in staging or something

no, it would be merged into master first. see the [flow of merged pull requests](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#flow-of-merged-pull-requests) section of CONTRIBUTING.md

> How do I add staging as an input using flakes and use linux-firmware from staging

since it's not in staging, you will need to use fpletz's fork

```nix
{
  inputs.nixpkgs.url = "nixpkgs/nixos-unstable";
  inputs.nixpkgs-firmware-fixed.url = "github:fpletz/nixpkgs/pkgs/compress-firmware-xz-ln-filesuffix";
  
  outputs = { nixpkgs, nixpkgs-firmware-fixed, ... }: {
    nixosConfigurations.myHost = nixpkgs.lib.nixosSystem {
      modules  = [
        ./configuration.nix
        {
          nixpkgs.overlays = [
            (_: prev: {
              inherit (nixpkgs-firmware-fixed.legacyPackages.${prev.stdenv.hostPlatform.system}) linux-firmware;
            })
          ];
        }
      ];
    };
  };
}
```

in the future i would also consider being more patient with merges, 2 days isn't that long :p

**Comment by beh-10257** [comment]
> > is this in staging or something
> 
> no, it would be merged into master first. see the [flow of merged pull requests](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#flow-of-merged-pull-requests) section of CONTRIBUTING.md
> 
> > How do I add staging as an input using flakes and use linux-firmware from staging
> 
> since it's not in staging, you will need to use fpletz's fork
> 
> ```nix
> {
>   inputs.nixpkgs.url = "nixpkgs/nixos-unstable";
>   inputs.nixpkgs-firmware-fixed.url = "github:fpletz/nixpkgs/pkgs/compress-firmware-xz-ln-filesuffix";
>   
>   outputs = { nixpkgs, nixpkgs-firmware-fixed, ... }: {
>     nixosConfigurations.myHost = nixpkgs.lib.nixosSystem {
>       modules  = [
>         ./configuration.nix
>         {
>           nixpkgs.overlays = [
>             (_: prev: {
>               inherit (nixpkgs-firmware-fixed.legacyPackages.${prev.stdenv.hostPlatform.system}) linux-firmware;
>             })
>           ];
>         }
>       ];
>     };
>   };
> }
> ```
> 
> in the future i would also consider being more patient with merges, 2 days isn't that long :p

@getchoo 
well since that doesn't work (there are no errors its just that the files haven't changed)
I guess this package doesn't change the linux-firmware package but yeah sorry for not waiting
and I don't know learning how to do this kind of stuff makes it so that I can use it just like I did with arch
anyway thanks anyway

**Comment by fpletz** [comment]
@getchoo Your overlay is actually not correct since this PR changes `compressFirmwareXz` that is used by the NixOS module merging the firmware packages and not `linux-firmware` itself.

Merging now myself since this PR got the approval of the derivation's author.

**Comment by NickCao** [comment]
This seems to break the loading of some amdgpu firmware, I'm still investigating.

**Comment by NickCao** [comment]
The offending dmesg line is:
```
Direct firmware load for amdgpu/psp_13_0_4_toc.bin failed with error -2
```

Before this change:
```
   <firmware>/lib/firmware/amdgpu/psp_13_0_4_toc.bin.xz
-> <linux-firmware-20240115-xz>/lib/firmware/amdgpu/psp_13_0_4_toc.bin.xz
-> psp_13_0_11_toc.bin.xz
```

After this change:
```
   <firmware>/lib/firmware/amdgpu/psp_13_0_4_toc.bin.xz does not exist
   <firmware>lib/firmware/amdgpu/psp_13_0_4_toc.bin
-> <linux-firmware-20240115-xz>/lib/firmware/amdgpu/psp_13_0_4_toc.bin
-> psp_13_0_11_toc.bin
```

**Comment by NickCao** [comment]
I see no easy fix other than reversing the logic to check for directories instead.

**Comment by spiage** [comment]
@NickCao https://github.com/NixOS/nixpkgs/commit/0dfe2e933f4418afdff7e69d1782b1aa86ca8b95#commitcomment-137937208

**Comment by corngood** [comment]
This broke my amdgpu firmware as well.  We could probably fix it like this:

```
diff --git a/pkgs/build-support/kernel/compress-firmware-xz.nix b/pkgs/build-support/kernel/compress-firmware-xz.nix
index cfb06a5c0f15..1c8e4df82a2e 100644
--- a/pkgs/build-support/kernel/compress-firmware-xz.nix
+++ b/pkgs/build-support/kernel/compress-firmware-xz.nix
@@ -15,6 +15,12 @@ runCommand "${firmware.name}-xz" args ''
           sh -c 'xz -9c -T1 -C crc32 --lzma2=dict=2MiB "${firmware}/$1" > "$1.xz"' --)
   (cd ${firmware} && find lib/firmware -type l) | while read link; do
       target="$(readlink "${firmware}/$link")"
-      ln -vs -- "''${target/^${firmware}/$out}.xz" "$out/$link.xz"
+      if [ -f "$(readlink -f "${firmware}/$link")" ]; then
+        ln -vs -- "''${target/^${firmware}/$out}.xz" "$out/$link.xz"
+      else
+        ln -vs -- "''${target/^${firmware}/$out}" "$out/$link"
+      fi
   done
+  # find broken symlinks
+  ! find $out -xtype l | grep .
 ''
```

bonus check for broken symlinks which fails both before (nvidia gsp directory links) and after (firmware bin links) the PR change.

**Comment by K900** [comment]
I tried rewriting it in Python: https://github.com/NixOS/nixpkgs/pull/284473

**Comment by K900** [comment]
The Python version notably produces no broken symlinks.

**Pull request merged**
